### PR TITLE
fix: serialize Dial/Listen queue handoff to prevent router token loss

### LINKED ISSUE #572: Flaky E2E: listen queue race causes 'Connection to exporter lost' in Dial/Listen handoff
## Summary

The `can lease and connect to exporters` E2E test fails intermittently on main with `Error: Connection to exporter lost`. The root cause is a race condition in the `listenQueues` sync.Map in `controller_service.go` that can lose router tokens during the Dial/Listen handoff.

## Symptoms

- Test `Core E2E Tests > Lease and connect > can lease and connect to exporters` fails at `e2e/test/e2e_test.go:406`
- Controller logs show repeated `Exporter in Available status, waiting for lease setup` retries (up to 9 attempts over 30 seconds)
- Exporter never transitions from Available to LeaseReady
- Client eventually gives up: `Error: Connection to exporter lost`

## Evidence on main

- **2026-04-15** upstream run `24468864095` failed with this exact pattern
- **2026-04-16** fork stress test run `24512850927` reproduced it (job `71649063409`)
- The previous fix attempt (commit `f473ede1`, "Increase Dial retry timeout from 10s to 30s") only widened the retry window but does not address the underlying race

## Root cause

`Listen()` and `Dial()` in `controller/internal/service/controller_service.go` both use `sync.Map.LoadOrStore` on `listenQueues` to share a buffered channel keyed by lease name. There are two race scenarios:

### Scenario 1: Two readers, one channel

When the exporter's Listen gRPC stream disconnects and reconnects, the new `Listen()` call hits `LoadOrStore` which returns the **existing** channel (the old entry was never removed). Both the old goroutine (still exiting) and the new goroutine read from the same channel. When Dial sends a router token, the old goroutine can consume it, attempt `stream.Send()` on its dead gRPC stream, get an error, and discard the token. The new goroutine never sees it.

### Scenario 2: Orphaned queue

If the old `Listen()` goroutine exits before the new one connects, the channel sits in the map with no active reader. `Dial()` sends a token into the buffered channel successfully (no blocking), but nothing ever drains it.

In both cases the exporter never receives the router token, never establishes a connection, and stays in `Available` status indefinitely.

### Vulnerable code

```go
// Listen() -- line 442
queue, _ := s.listenQueues.LoadOrStore(leaseName, make(chan *pb.ListenResponse, 8))
for {
    select {
    case <-ctx.Done():
        return nil  // exits without removing queue from map
    case msg := <-queue.(chan *pb.ListenResponse):
        if err := stream.Send(msg); err != nil {
            return err  // exits without removing queue from map
        }
    }
}

// Dial() -- line 735
queue, _ := s.listenQueues.LoadOrStore(leaseName, make(chan *pb.ListenResponse, 8))
select {
case <-ctx.Done():
    return nil, ctx.Err()
case queue.(chan *pb.ListenResponse) <- response:  // sends to potentially stale queue
}
```

The fundamental issue: `Listen()` never cleans up its queue entry on exit, and there is no mechanism to detect or prevent a stale reader from consuming tokens.

## Reproduction

A unit test confirms this is not a rare edge case. With two goroutines reading from the same channel (modeling old + new Listen), the wrong goroutine consumes the token ~50% of the time:

```
=== RUN   TestListenQueueRace_OverlapStress
    old goroutine stole the token: 2510 / 5000 (50.2%)
    new goroutine got the token:   2490 / 5000 (49.8%)
    RACE CONFIRMED: in 50.2% of iterations the dying Listen goroutine
    consumed the Dial token before the live one could
--- PASS: TestListenQueueRace_OverlapStress (3.78s)
```

The test file is at `controller/internal/service/listen_queue_race_test.go` on the `fix-listen-queue-race` branch.

## Related

- PR #417 addresses this with per-lease mutex serialization, done channels, and non-blocking sends
- Issue #538 / commit `f473ede1` increased the Dial retry timeout as a workaround but does not fix the race

### PR DESCRIPTION
**Pull request by raballew**
## Summary

- Fixes the TOCTOU race in `listenQueues` where a reconnecting `Listen()` goroutine and a dying one compete for the same channel, causing ~50% token loss rate
- Wraps the raw channel in a `listenQueue` struct with a `done` channel and `sync.Once` for clean shutdown signaling
- Introduces a per-lease mutex (`leaseLock`) that serializes queue swaps in `Listen()` and token sends in `Dial()`, eliminating the race window entirely
- Uses non-blocking sends to prevent mutex starvation when the channel buffer is full
- Adds ref-counted lease lock cleanup to prevent memory leaks in the `leaseLocks` sync.Map
- Enables the race detector in the controller test target

## Test plan

- [ ] `go test -race -v ./internal/service/ -run TestListenQueue` passes
- [ ] `make e2e-run` passes with label filter `core`
- [ ] Stress test workflow confirms reduced flake rate

Fixes #572

Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by raballew** [comment]
@mangelajo ptal

**Comment by raballew** [comment]
@mangelajo bumping this one. needs review.

**Review by mangelajo — APPROVED** [review]
now an abridged version :-P :)

File: controller/internal/service/controller_service.go [hunk 1]
```diff
@@ -80,6 +81,113 @@ type ControllerService struct {
...
+func (s *ControllerService) releaseLeaseLock(leaseName string) {
+	v, ok := s.leaseLocks.Load(leaseName)
+	if !ok {
+		return
+	}
+	ll := v.(*leaseLock)
+	if atomic.AddInt32(&ll.refs, -1) == 0 {
+		s.leaseLocks.CompareAndDelete(leaseName, ll)
+	}
+}
```
[the remark this answers is not in the document]
Comment (mangelajo) [reply controller/internal/service/controller_service.go:138]: No, CodeRabbit's suggestion does not make sense -- it's a false positive. The bot failed to reason through the existing retry logic correctly.

CodeRabbit's claim
acquireLeaseLock can hand out a mutex that another goroutine removes from the map in releaseLeaseLock, causing different callers to get different mutexes for the same leaseName

Why it's wrong
The newRefs <= 1 check already prevents this. Let me trace through the exact scenario CodeRabbit is worried about:

```
Time   releaseLeaseLock (G1)              acquireLeaseLock (G2)
────   ──────────────────────             ─────────────────────
 1     atomic.Add(&refs, -1) → 0
                                           LoadOrStore → gets old ll (still in map)
 2                                         atomic.Add(&refs, +1) → 1
                                           newRefs=1, which is <= 1 → BACKS OUT
                                           atomic.Add(&refs, -1) → 0
                                           continue (retry loop!)
 3     CompareAndDelete → removes old ll
 4                                         LoadOrStore → creates fresh ll2 (refs=1)
                                           loaded=false → returns &ll2.mu  ✓
G2 never returns the stale mutex. The newRefs <= 1 check catches it every time.
```

The only way acquireLeaseLock returns a mutex is when newRefs > 1, which means refs was >= 1 before our increment -- i.e., at least one other goroutine is actively holding a reference and hasn't released yet. Since our increment brings it to N+1, all N+1 holders would need to release before refs hits 0. Since we haven't released, the entry cannot be torn down while we hold it.

Why CodeRabbit's proposed fix is worse
It replaces a lock-free, per-lease mechanism with a global mutex (leaseLocksMu) that serializes all lease lock operations across every lease. This adds unnecessary contention -- every Dial and every Listen across all leases would contend on the same mutex just to acquire/release their per-lease lock. The current approach correctly avoids that.

I'd dismiss this comment as a false positive.
